datalad/local/copy_file.py

```diff
@@ -0,0 +1,665 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Copy files (from one dataset to another)"""
+
+__docformat__ = 'restructuredtext'
+
+
+import logging
+import os.path as op
+from shutil import copyfile
+import sys
+
+from datalad.interface.base import Interface
+from datalad.interface.utils import eval_results
+from datalad.interface.base import build_doc
+from datalad.support.constraints import (
+    EnsureStr,
+    EnsureNone,
+)
+from datalad.interface.common_opts import save_message_opt
+from datalad.support.param import Parameter
+from datalad.distribution.dataset import (
+    Dataset,
+    require_dataset,
+)
+from datalad.support.annexrepo import AnnexRepo
+from datalad.utils import (
+    assure_list,
+    get_dataset_root,
+    Path,
+)
+
+from datalad.distribution.dataset import (
+    EnsureDataset,
+    datasetmethod,
+    resolve_path,
+)
+
+lgr = logging.getLogger('datalad.local.copy_file')
+
+
+@build_doc
+class CopyFile(Interface):
+    """Copy files and their availability metadata from one dataset to another.
+
+    The difference to a system copy command is that here additional content
+    availability information, such as registered URLs, is also copied to the
+    target dataset. Moreover, potentially required git-annex special remote
+    configurations are detected in a source dataset and are applied to a target
+    dataset in an analogous fashion. It is possible to copy a file for which no
+    content is available locally, by just copying the required metadata on
+    content identity and availability.
+
+    || REFLOW >>
+    Interface and behavior are modeled after the POSIX 'cp' command, with two
+    main differences. 1) Via [CMD: --specs-from CMD][PY: `specs_from` PY] it is
+    possible to flexibly input source-destination path pairs. 2) Saving changes
+    in a target dataset is conditional upon a specification of
+    [CMD: --dataset CMD][PY: `dataset`, or calling this command as a `Dataset`
+    method PY].
+    << REFLOW ||
```

Comment:
The added docstring overall looks good to me, but I don't like this section.  I think "interface *and behavior*" is a bit too broad for the reader to know what you have in mind when you're listing these differences, and how they, er, differ from the differences the last paragraph just mentioned.  Perhaps something like

    The interface is modeled after the POSIX 'cp' command, but with
    one additional way to specify what to copy where: [CMD:
    --specs-from CMD][PY: `specs_from` PY] allows the caller to
    flexibly input source-destination path pairs.

That mostly just drops the `--dataset` bit.  I think it was included because it's the other copy-file argument that cp doesn't have.  But that's obvious enough that I don't think it's worth an explicit mention, especially because doing so would mean that when `copy-file` gains a new option, this part would need to be updated.

Reply:
I can follow your logic. Applied!